src/vma/dev/ring_tap.cpp

```diff
@@ -0,0 +1,1213 @@
+		}
+		ring_logerr("Transport type is ETH, dropping the packet");
+		return false;
+	}
+#endif
+	break;
+
+	default:
+		ring_logwarn("Rx packet dropped - undefined protocol = %d", p_ip_h->protocol);
+		return false;
+	}
+
+	if (unlikely(p_rfs == NULL)) {
+		ring_logdbg("Rx packet dropped - rfs object not found: dst:%d.%d.%d.%d:%d, src%d.%d.%d.%d:%d, proto=%s[%d]",
+				NIPQUAD(p_rx_wc_buf_desc->rx.dst.sin_addr.s_addr), ntohs(p_rx_wc_buf_desc->rx.dst.sin_port),
+				NIPQUAD(p_rx_wc_buf_desc->rx.src.sin_addr.s_addr), ntohs(p_rx_wc_buf_desc->rx.src.sin_port),
+				iphdr_protocol_type_to_str(p_ip_h->protocol), p_ip_h->protocol);
+
+		return false;
+	}
+	return p_rfs->rx_dispatch_packet(p_rx_wc_buf_desc, pv_fd_ready_array);
+}
+
+void ring_tap::send_ring_buffer(ring_user_id_t id, vma_ibv_send_wr* p_send_wqe, vma_wr_tx_packet_attr attr)
+{
+	NOT_IN_USE(id);
+	auto_unlocker lock(m_lock_ring_tx);
+	int ret = send_buffer(p_send_wqe, attr);
+	send_status_handler(ret, p_send_wqe);
+}
+
+void ring_tap::send_lwip_buffer(ring_user_id_t id, vma_ibv_send_wr* p_send_wqe, bool b_block)
+{
+	NOT_IN_USE(id);
+	auto_unlocker lock(m_lock_ring_tx);
+	mem_buf_desc_t* p_mem_buf_desc = (mem_buf_desc_t*)(p_send_wqe->wr_id);
+	p_mem_buf_desc->lwip_pbuf.pbuf.ref++;
+	vma_wr_tx_packet_attr attr = (vma_wr_tx_packet_attr)((b_block*VMA_TX_PACKET_BLOCK)|VMA_TX_PACKET_L3_CSUM|VMA_TX_PACKET_L4_CSUM);
+	int ret = send_buffer(p_send_wqe, attr);
+	send_status_handler(ret, p_send_wqe);
+}
+
+void ring_tap::prepare_flow_message(vma_msg_flow& data,
+		flow_tuple& flow_spec_5t, msg_flow_t flow_action)
+{
+	memset(&data, 0, sizeof(data));
+	data.hdr.code = VMA_MSG_FLOW;
+	data.hdr.ver = VMA_AGENT_VER;
+	data.hdr.pid = getpid();
+	data.action = flow_action;
+	data.if_id = get_parent()->get_if_index();
+	data.tap_id = get_if_index();
+	if (flow_spec_5t.is_3_tuple()) {
+		data.type = VMA_MSG_FLOW_TCP_3T;
+		data.flow.t3.dst_ip = flow_spec_5t.get_dst_ip();
+		data.flow.t3.dst_port = flow_spec_5t.get_dst_port();
+	} else {
+		data.type = VMA_MSG_FLOW_TCP_5T;
+		data.flow.t5.src_ip = flow_spec_5t.get_src_ip();
+		data.flow.t5.src_port = flow_spec_5t.get_src_port();
+		data.flow.t5.dst_ip = flow_spec_5t.get_dst_ip();
+		data.flow.t5.dst_port = flow_spec_5t.get_dst_port();
+	}
+}
+
+int ring_tap::process_element_rx(void* pv_fd_ready_array)
+{
+	int ret = 0;
+
+	if(m_tap_data_available) {
+		auto_unlocker lock(m_lock_ring_rx);
+		if (m_rx_pool.size() || request_more_rx_buffers()) {
+			mem_buf_desc_t *buff = m_rx_pool.get_and_pop_front();
+			ret = orig_os_api.read(m_tap_fd, buff->p_buffer, buff->sz_buffer);
+			if (ret > 0) {
+				/* Data was read and processed successfully */
+				buff->sz_data = ret;
+				if (m_vf_ring) {
+					m_vf_ring->rx_process_buffer(buff, pv_fd_ready_array);
+				} else {
+					rx_process_buffer(buff, pv_fd_ready_array);
+				}
+				m_p_ring_stat->n_rx_byte_count += ret;
+				m_p_ring_stat->n_rx_pkt_count++;
+				m_p_ring_stat->tap.n_rx_buffers--;
+			} else {
```

Comment:
remove stats update, why processing using the vf ring ? we should enable sw_csum in this case.
readv ?